docker/Dockerfile

```diff
@@ -1,49 +1,56 @@
 # Production Docker backend image definition
 # Used in CI for creating images to push on registry
 
-# Update the VARIANT arg in docker-compose.yml to pick a Node version: 10, 12, 14
-ARG VARIANT=20
-FROM node:${VARIANT}
-
-# Update args in docker-compose.yaml to set the UID/GID of the "node" user.
-ARG USER_UID=1000
-ARG USER_GID=$USER_UID
-RUN if [ "$USER_GID" != "1000" ] || [ "$USER_UID" != "1000" ]; then \
-  groupmod --gid $USER_GID node \
-  && usermod --uid $USER_UID --gid $USER_GID node \
-  && chown -R $USER_UID:$USER_GID /home/node \
-  && chown -R $USER_UID:root /usr/local/share/nvm /usr/local/share/npm-global; \
-  fi
-
-# variable passed as a build arg. Represents the tag or git sha used for the build
-ARG APP_VERSION
-
-# [Optional] Uncomment this section to install additional OS packages.
-# RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
-#     && apt-get -y install --no-install-recommends <your-package-list-here>
-
-# [Optional] Uncomment if you want to install an additional version of node using nvm
-# ARG EXTRA_NODE_VERSION=10
-# RUN su node -c "source /usr/local/share/nvm/nvm.sh && nvm install ${EXTRA_NODE_VERSION}"
-
-# [Optional] Uncomment if you want to install more global node packages
-# RUN sudo -u node npm install -g <your-package-list-here>
+# Update the NODE_VERSION arg in docker-compose.yml to pick a Node version of Docker Image
+ARG NODE_VERSION=20
 
-USER node
-
-WORKDIR /home/node
-
-RUN mkdir /home/node/app
-RUN mkdir /home/node/file-storage
+# Update the ALPINE_VERSION arg in docker-compose.yml to pick a Node version of Docker Image
+ARG ALPINE_VERSION=3.19
 
-COPY --chown=$USER_UID:$USER_GID . /home/node/app
+# Variable passed as a build arg. Represents the tag or git sha used for the build
+ARG APP_VERSION 
 
-WORKDIR /home/node/app
+##############################
+### Build Application Step ###
+##############################
+# This step will build the application
+FROM node:${NODE_VERSION} AS builder
 
-RUN yarn install
+WORKDIR /app
+COPY . .
+RUN yarn install --ignore-scripts
 
 RUN yarn build-ts
 
-# set APP_VERSION as ENV variable from ARG passed at build step
+############################
+### Run Application Step ###
+############################
+# Final step that will run the application
+FROM node:${NODE_VERSION}-alpine${ALPINE_VERSION} as runner
+
+# Set APP_VERSION as ENV variable from ARG passed at build step
 ENV APP_VERSION=${APP_VERSION:-latest}
-CMD NODE_PATH=./dist node dist/server.js
+
+EXPOSE 3000
+
+WORKDIR /app
+
+# Add needed packages
+RUN apk update \
+    && apk upgrade \
+    && apk add \
+        yarn \
+    # Python and GCC are necessary for mmmagic dependency.
+        python3 \
+        py3-pip \
+        make \
+        build-base
+
+COPY ./package.json ./
+
+# Install the application
+RUN yarn install --production --ignore-scripts
+COPY --from=builder ./app/dist ./dist
+
+USER node
+CMD yarn start
```

Comment:
Is it better practice to do `yarn start` than to use `node` on the built server.js ?

Reply:
Ah I see that `yarn start` is executing `yarn serve` which does exactly `NODE_PATH=./dist node dist/server.js`